modules/images/regenerate-existing-images/background-process/class-perflab-background-job.php

```diff
@@ -0,0 +1,403 @@
+			self::JOB_STATUS_QUEUED,
+			self::JOB_STATUS_RUNNING,
+			self::JOB_STATUS_PARTIAL,
+		);
+
+		if ( in_array( $status, $valid_statuses, true ) ) {
+			update_term_meta( $this->job_id, self::JOB_STATUS_META_KEY, $status );
+
+			// If job is complete, set the timestamp at which it was completed.
+			if ( self::JOB_STATUS_COMPLETE === $status ) {
+				update_term_meta( $this->job_id, 'job_completed_at', time() );
+			}
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Records the error for the current process run.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param WP_Error $error Error instance.
+	 *
+	 * @todo Implement the error recording activity.
+	 *
+	 * @return void
+	 */
+	public function set_error( WP_Error $error ) {
+		$job_failure_data = $error->get_error_data( 'perflab_job_failure' );
+
+		if ( ! empty( $job_failure_data ) ) {
+			$this->set_status( self::JOB_STATUS_FAILED );
+
+			update_term_meta( $this->job_id, self::JOB_ERRORS_META_KEY, $job_failure_data );
+			update_term_meta( $this->job_id, self::JOB_ATTEMPTS_META_KEY, ( $this->get_attempts() + 1 ) );
+		}
+	}
+
+	/**
+	 * Retrieves the job status from its meta information.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string Job status.
+	 */
+	public function get_status() {
+		return (string) get_term_meta( $this->job_id, self::JOB_STATUS_META_KEY, true );
+	}
+
+	/**
+	 * Retrieves the job data.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array|null Job data.
+	 */
+	public function get_data() {
+		// If we have cached data, return it.
+		if ( ! is_null( $this->data ) ) {
+			return $this->data;
+		}
+
+		$this->data = get_term_meta( $this->job_id, self::JOB_DATA_META_KEY, true );
+
+		return $this->data;
+	}
+
+	/**
+	 * Retrieves the job name.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string Job name.
+	 */
+	public function get_name() {
+		return (string) $this->name;
+	}
+
+	/**
+	 * Creates a new job in the queue.
+	 *
+	 * Job refers to the term and queue refers to the `background_job` taxonomy.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $name Name of job identifier.
+	 * @param array  $data Data for the job.
+	 *
+	 * @return Perflab_Background_Job|WP_Error Job object if created successfully, else WP_Error.
+	 */
+	public static function create( $name, array $data = array() ) {
+		// Insert the new job in queue.
+		$term_data = wp_insert_term( 'job_' . time(), PERFLAB_BACKGROUND_JOB_TAXONOMY_SLUG );
+
+		if ( ! is_wp_error( $term_data ) ) {
+			update_term_meta( $term_data['term_id'], self::JOB_DATA_META_KEY, $data );
+			update_term_meta( $term_data['term_id'], self::JOB_NAME_META_KEY, $name );
+
+			/**
+			 * Create a fresh instance to return.
+			 *
+			 * @var Perflab_Background_Job
+			 */
+			$job = new self( $term_data['term_id'] );
+
+			$job->name = $name;
+			$job->data = $data;
+
+			// Set the queued status for freshly created jobs.
+			$job->set_status( self::JOB_STATUS_QUEUED );
+
+			return $job;
+		}
+
+		return $term_data;
+	}
```

Comment:
Instead of a static method here, we should rather introduce this as a function, to follow core's typical pattern. The job class here should be mostly for read access for a specific job, but infrastructure pieces e.g. to create, update, or delete a job should rather be implemented as functions.

I would suggest we move this into a `perflab_insert_background_job()` function.

Reply:
As of now, there is no case to update the job. So I will add create and delete functions.